proof_of_concept/registry.py

```diff
@@ -38,115 +38,61 @@ def __init__(self, name: str, public_key: RSAPublicKey) -> None:
 
         Args:
             name: Name of the site.
+            owner: Party which owns this site.
             admin: Party which administrates this site.
+            runner: This site's local workflow runner.
+            store: This site's asset store.
+            namespace: The namespace managed by this site's policy
+                server.
+            policy_server: This site's policy server.
 
         """
         self.name = name
+        self.owner = owner
         self.admin = admin
-
-
-class RunnerDescription(RegisteredObject):
-    """Describes a workflow runner to the rest of the DDM.
-
-    Attributes:
-        site: Site at which this runner is located.
-        runner: The runner service.
-
-    """
-    def __init__(
-            self, site: SiteDescription, runner: ILocalWorkflowRunner
-            ) -> None:
-        """Create a RunnerDescription.
-
-        Args:
-            site: Site at which this runner is located.
-            runner: The runner service.
-
-        """
-        self.site = site
         self.runner = runner
-
-
-class AssetStoreDescription(RegisteredObject):
-    """Describes an asset store to the rest of the DDM.
-
-    Attributes:
-        site: The site at which this store is located.
-        store: The store service.
-
-    """
-    def __init__(
-            self, site: SiteDescription, store: IAssetStore
-            ) -> None:
-        """Create an AssetStoreDescription.
-
-        Args:
-            site: The site at which this store is located.
-            store: The store service.
-
-        """
-        self.site = site
         self.store = store
+        self.namespace = namespace
+        self.policy_server = policy_server
 
+        if store is None and runner is not None:
+            raise RuntimeError('Site with runner needs a store')
 
-_ReplicatedClass = TypeVar('_ReplicatedClass', bound=RegisteredObject)
-
-
-class PolicyServerDescription(RegisteredObject):
-    """Describes a policy server to the rest of the DDM.
+        if namespace is None and policy_server is not None:
+            raise RuntimeError('Policy server specified without namespace')
 
-    Attributes:
-        namespace: The namespace governed by this server.
-        site: The site at which this server is located.
+        if namespace is not None and policy_server is None:
+            raise RuntimeError('Namespace specified but policy server missing')
 
-    """
-    def __init__(
-            self, site: SiteDescription, namespace: NamespaceDescription,
-            server: IPolicyServer) -> None:
-        """Create a PolicyServerDescription.
 
-        Args:
-            site: The site at which this server is located.
-            namespace: The namespace governed by this server.
-            server: The policy server.
-
-        """
-        self.site = site
-        self.namespace = namespace
-        self.server = server
+_ReplicatedClass = TypeVar('_ReplicatedClass', bound=RegisteredObject)
```

Comment:
Maybe factor this out into a method `_validate()`

Reply:
Additionally, I was wondering whether this validation lives here, and not under Site construction or when registering a site. (naive question)

Reply:
It's not factoring out if there's only one occurrence :-). It would add two extra lines of code and an extra function, with the only benefit that you make it explicit that this is validating the object, but I think that's clear enough as is.

The reason I'm checking this here is that it simply makes no sense to have a namespace without a policy server for it or vice versa. So this is a class invariant of `SiteDescription`, and those should be checked in the constructor to ensure that there are no invalid instances around. Of course input validation should be done in any REST API version of  this, but that's a separate issue I think.

Come to think of it, we could actually have the user submit a `SiteDescription` object in `Registry.register_site()`, instead of a bunch of separate fields. I'd like to remove the `ReplicatedObject` base class then though, since it's an implementation detail of the Registry and should not be in a class used outside of it. It did occur to me some time ago that you can probably use a Union rather than having a base class, so that should be possible. I'll have a look; maybe it makes sense to do the same for parties then as well, and it may provide a cleaner base to start the REST API from.

As for `Site`, that's a separate class which represents the actual site, where `SiteDescription` represents a digital document in the Registry which describes a site. The `Site` constructor creates the policy server, so it should be correct by construction.

Reply:
Making it more explicit then (I would still call that factoring out). I guess this is very subjective though so feel free to leave it like this! 👍 

Indeed submitting a `SiteDescription` object to `register_site` makes a lot of sense!

And regarding `Site`, indeed the current constructor produces a valid `Site` so no need to validate.

Reply:
I'm not sure if the term really comes from there, but I think about the algebraic transformation `a * x + a * y => a * (x + y)`, where there's a common factor `a` that is taken out of the equation and moved to the front, removing duplication. I'm not sure whether Python code is an algebraic ring however :-).

I've moved the `PartyDescription` and `SiteDescription` to definitions.py now, and use them when registering. This also changes `SiteDescription` to refer to the owner and the admin by name, as opposed to having a pointer to an object. That needs to happen anyway for serialisation, and it only adds one lookup in `DDMClient`, so that works out fine.

Reply:
Sounds good! 👍